fetch/api/resources/redirect.py

```diff
@@ -2,68 +2,70 @@
-               ("Cache-Control", "no-cache"),
-               ("Pragma", "no-cache")]
-    if "Origin" in request.headers:
-        headers.append(("Access-Control-Allow-Origin", request.headers.get("Origin", "")))
-        headers.append(("Access-Control-Allow-Credentials", "true"))
+    headers = [(b"Content-Type", b"text/plain"),
+               (b"Cache-Control", b"no-cache"),
+               (b"Pragma", b"no-cache")]
+    if b"Origin" in request.headers:
+        headers.append((b"Access-Control-Allow-Origin", request.headers.get(b"Origin", b"")))
+        headers.append((b"Access-Control-Allow-Credentials", b"true"))
     else:
-        headers.append(("Access-Control-Allow-Origin", "*"))
+        headers.append((b"Access-Control-Allow-Origin", b"*"))
 
     token = None
-    if "token" in request.GET:
-        token = request.GET.first("token")
+    if b"token" in request.GET:
+        token = request.GET.first(b"token")
         data = request.server.stash.take(token)
         if data:
             stashed_data = data
 
-    if request.method == "OPTIONS":
-        if "allow_headers" in request.GET:
-            headers.append(("Access-Control-Allow-Headers", request.GET['allow_headers']))
-        stashed_data['preflight'] = "1"
+    if request.method == u"OPTIONS":
+        if b"allow_headers" in request.GET:
+            headers.append((b"Access-Control-Allow-Headers", request.GET[b'allow_headers']))
+        stashed_data[b'preflight'] = b"1"
         #Preflight is not redirected: return 200
-        if not "redirect_preflight" in request.GET:
+        if not b"redirect_preflight" in request.GET:
             if token:
-                request.server.stash.put(request.GET.first("token"), stashed_data)
-            return 200, headers, ""
+                request.server.stash.put(request.GET.first(b"token"), stashed_data)
+            return 200, headers, u""
 
-    if "redirect_status" in request.GET:
-        status = int(request.GET['redirect_status'])
+    if b"redirect_status" in request.GET:
+        status = int(request.GET[b'redirect_status'])
 
-    stashed_data['count'] += 1
+    stashed_data[b'count'] += 1
 
-    if "location" in request.GET:
-        url = request.GET['location']
-        if "simple" not in request.GET:
+    if b"location" in request.GET:
+        url = isomorphic_decode(request.GET[b'location'])
+        if b"simple" not in request.GET:
             scheme = urlparse(url).scheme
-            if scheme == "" or scheme == "http" or scheme == "https":
-                url += "&" if '?' in url else "?"
+            if scheme == u"" or scheme == u"http" or scheme == u"https":
+                url += u"&" if u'?' in url else u"?"
                 #keep url parameters in location
                 url_parameters = {}
                 for item in request.GET.items():
                     url_parameters[item[0]] = item[1][0]
                 url += urlencode(url_parameters)
```

Comment:
I wonder if urlencode gets messed by the byte strings from request.GET.items?

Reply:
Python 3 (from version 3.2) urlencode query parameter supports bytes and string objects. I did a few printouts and the output looked reasonable.